[WIP] Read OpenSSH server configuration and warn about change of default in PermitRootLogin

Jakuje opened the pull request:
This PR creates separate actors, where one knows how to read configuration file and the other acts on one option read from this file.

It is still WIP and subject to improvements in coming days.

Jakuje commented:
Note, that reading `sshd_config` requires the process to run under root. I was not able to find a documentation to this whether this can be indicated somehow or whether this is automatic.

Jakuje commented:
@vojtechsokol This is what I came up with during last days to cover pre-upgrade checks in OpenSSH, which might cause issues during updating RHEL7 based on the previous document. Please, let me know whether you notice some issues, or what can be improved.

There are still some TODOs and questions that will probably need answers, clarifications or improvements.

vojtechsokol commented on repos/system_upgrade/el7toel8/actors/readopensshconfig/libraries/readopensshconfig.py:
```diff
@@ -0,0 +1,93 @@
+
+
+def scan_sshd(producer):
+    ''' Parse sshd_config configuration file to create OpenSshConfig message.
```
Malformed docstring - use double quotes, closing quotes are on the same line as the opening ones for one-line docstring, see the [guidelines](https://github.com/oamg/leapp-repository/blob/master/CONTRIBUTING.md#code-style).

vojtechsokol commented on repos/system_upgrade/el7toel8/models/opensshconfig.py:
```diff
@@ -0,0 +1,21 @@
+
+
+class OpenSshConfig(Model):
+    ''' This model does not actually contain effective configuration options
```
Malformed docstring - use double quotes, one-line summary (phrase ending in a period) + blank line + extended description for multi-line docstring, see the [guidelines](https://github.com/oamg/leapp-repository/blob/master/CONTRIBUTING.md#code-style).

Jakuje commented:
Can the PEP8 linter be automatically running on the PRs so it can save time to both contributors and reviewers, not having to mention marginal things and focus on the actual logic and functionality? Sorry, but it is some time since I wrote some larger chunk of code in python.

I addressed your comments and rebased the branch. About the todos:

* `actors/opensshpermitrootlogin/tests/test_library.py`:    # TODO This is suspicious configuration we should probably handle separately
  * This is something I would like to address, but the question emerged in my head whether there is some sensible way how this can be figured out globally or whether there are some other models that could give me this information (from rpm), or whether I should just call `rpm -qVf /etc/ssh/sshd_config`.
* `actors/opensshpermitrootlogin/tests/test_library.py`:    # TODO This is suspicious configuration we should probably handle separately
  * This is something that is probably not very important enough to care for. This configuration is actually changing semantics of the default block after update, but there is still some condition where the application is still accessible from.

vojtechsokol commented:
- About the linter
   There is a plan to add linter to our CI setup, however it was put aside due to the GA rush. Also it was discussed that maybe at least part of the pipeline will be done in Travis CI to make it visible to anyone, and adding linter to Travis is one liner, so hopefully soon...
- About the first TODO
    I suppose this relates to the two TODO comments in [OpenSshPermitRootLogin](https://github.com/oamg/leapp-repository/pull/184/files#diff-8917fe4eded1b245f2d8e00b88cb85ddR23) and [ReadOpenSshConfig](https://github.com/oamg/leapp-repository/pull/184/files#diff-50028f6f03dfd7ee068f20cb54bfaeadR89) actors. I am not aware of any means how can it be done besides the mentioned `rpm` command. But IIRC `rpm -Vf` gives you only a list of changed files, not the actual diff. For that, you'll need to run `rpmpeek` or some similar tool on the downloaded rpm package.
- About the second TODO
    If it's not important then I think it's OK to just leave a TODO comment in the code as it is right now.

Jakuje commented:
> But IIRC rpm -Vf gives you only a list of changed files, not the actual diff. For that, you'll need to run rpmpeek or some similar tool on the downloaded rpm package.

I do not actually need a diff, but I just need to know whether the update will overwrite the configuration file or not, which is, I think, when the checksum of configuration file matches.

I will rename the actors to match your recommendation to be more maintainable in the future.

One more thing about the level of reporting and the messages content, whether it is too many of them, not enough of them, too much information, too few or so.

Jakuje commented:
Rebased to reflect the above recommended naming, added the rpm -V parth and fixed few minor issues.

vojtechsokol commented on repos/system_upgrade/el7toel8/actors/opensshalgorithmscheck/actor.py:
```diff
@@ -0,0 +1,60 @@
+                        % ','.join(found_ciphers),
+                remediation='Remove the following ciphers from sshd_config: '
+                            '%s' % ','.join(found_ciphers),
+                severity='high')
```
Shouldn't this report inhibit the upgrade?

Jakuje replied:
Yes, you are right. These would prevent the sshd startup. Fixing in the next push.

vojtechsokol commented on repos/system_upgrade/el7toel8/actors/opensshalgorithmscheck/actor.py:
```diff
@@ -0,0 +1,60 @@
+                        % ','.join(found_macs),
+                remediation='Remove the following MACs from sshd_config: '
+                            '%s' % ','.join(found_macs),
+                severity='high')
```
Shouldn't this report inhibit the upgrade?

Jakuje replied:
Fixed also this one

Jakuje commented on repos/system_upgrade/el7toel8/actors/opensshalgorithmscheck/actor.py:
```diff
@@ -0,0 +1,60 @@
+                summary='OpenSSH is configured to use removed ciphers %s. '
+                        'These ciphers were removed from OpenSSH and if '
+                        'present the sshd daemon will not start in RHEL 8'
+                        % ','.join(found_ciphers),
```
Fixed throughout the code

Jakuje commented:
Fixed the last batch of comments and tried to accommodate some more alerts from pylint.

I do not think there will be much more checks in the future for OpenSSH, since I think I covered more than enough already (one missing thing is tcp wrappers after merging #186. The idea of separating them came that they can be simply removed when they will not be needed in the next update from el8 and the new actors can be created, allowing to keep them pretty simple and independent from each other. From what I see, with `PermitRootLogin`, we will have issues also during the next updates (el8->el9).

vojtechsokol commented:
Thanks for the contribution, merging.
